dev-tools/releng/src/job.rs

```diff
@@ -0,0 +1,284 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use std::process::Stdio;
+use std::sync::Arc;
+use std::time::Instant;
+
+use anyhow::anyhow;
+use anyhow::Context;
+use anyhow::Result;
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use fs_err::tokio::File;
+use futures::stream::FuturesUnordered;
+use futures::stream::TryStreamExt;
+use slog::info;
+use slog::Logger;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::AsyncRead;
+use tokio::io::AsyncWrite;
+use tokio::io::AsyncWriteExt;
+use tokio::io::BufReader;
+use tokio::process::Command;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::error::RecvError;
+use tokio::sync::Semaphore;
+
+use crate::cmd::CommandExt;
+
+// We want these two jobs to run without delay because they take the longest
+// amount of time, so we allow them to run without taking a permit first.
+const PERMIT_NOT_REQUIRED: [&str; 2] = ["host-package", "host-image"];
+
+pub(crate) struct Jobs {
+    logger: Logger,
+    permits: Arc<Semaphore>,
+    log_dir: Utf8PathBuf,
+    map: HashMap<String, Job>,
+}
+
+struct Job {
+    future: Pin<Box<dyn Future<Output = Result<()>>>>,
```

Comment:
1. Can this use https://docs.rs/futures/latest/futures/future/type.BoxFuture.html instead?
2. Should this be `Send`? (`BoxFuture` is `Send`.)

Reply:
It appears the answer to "should this be `Send`" is "yes, because I am about to give it to `tokio::spawn`. (This makes it equivalent to `BoxFuture` so I'll use that.)